R/build-tools.R

```diff
@@ -0,0 +1,39 @@
+#' Build Tools
+#'
+#' Check and install build tools as required.
+#'
+#' These functions are intended to be used together -- one should
+#' first check whether build tools are available, and when not,
+#' prompt for installation. For example:
+#'
+#' ```R
+#' compile_model <- function(...) {
+#'   if (!rstudioapi::buildToolsCheck())
+#'     rstudioapi::buildToolsInstall("Model compilation")
+#'
+#'   # proceed with compilation
+#' }
+#' ```
+#'
+#' The `action` parameter is used to communicate (with a prompt) the operation
+#' being performed that requires build tool installation. Setting it to `NULL`
+#' or the empty string will suppress that prompt.
+#'
+#' @param action The action (as a string) being taken that will require
+#'   installation of build tools.
+#'
+#' @note The `buildToolsCheck()` and `buildToolsInstall()` functions were added
+#'   with version 1.2.945 of RStudio.
+#'
+#' @name build-tools
+NULL
+
+#' @name build-tools
+buildToolsCheck <- function() {
+  callFun("buildToolsCheck")
```

Comment:
Does this cache it's result? If not this might be a bit expensive esp. on Windows.

We could also argue that it shouldn't cache at all or that `buildToolsInstall()` should invalidate the cache.

In any case, just wanted to put this concern on your radar.

Reply:
It does not cache the result, and running this does imply attempting to compile a dummy `.c` file with `R CMD SHLIB` so could indeed be expensive.

I can't think of a good reason why we wouldn't cache a successful check of the build tools check, though. I imagine it's unlikely that many users would deliberately remove build tools from their path; and if they did, restarting the session would clear that state anyhow.

Maybe it's fine to just cache this by default (when we have successfully detected compile tools)?